CHANGELOG.md

```diff
@@ -13,6 +13,10 @@ Breaking changes
   `AtomicSwapRecipientQuery`, `AtomicSwapSenderQuery`, `AtomicSwapIdQuery`,
   `AtomicSwapHashlockQuery`, `isAtomicSwap*Query`.
 * @iov/bcp-types: Rename `bnsSwapQueryTags` to `bnsSwapQueryTag`
+* @iov/bcp-types: Let `SwapOfferTransaction` take a `hash` instead of a `preimage`.
```

Comment:
Doing this means SwapOffer and SwapCounter are the same and can be combined.

I guess this means we need a standard preimage calculator exposed, which upon reflection is important. If I use a different hash algorithm on chain A and chain B, then I can claim this, while giving a preimage that is useless.

We should also standardize maximum allowed size of the preimage.

But that is a different issue...

Reply:
You are right, SwapOfferTransaction and SwapCounterTransaction contains the same fields (probably they are implemented as the same transaction in BNS). But maybe they'll diverge at some point? For me the differentiation right now helps to express the intent from a user's perspective.

> If I use a different hash algorithm on chain A and chain B, then I can claim this, while giving a preimage that is useless.

The hashing algorithm must be the same on both chains (for the reason you gave). Right now this algorithm is hardcoded to SHA256, which is kind of globally known. We could make it configurable probably but I see this as a pre-launch thing.

> We should also standardize maximum allowed size of the preimage.

Very valid and non-trivial point, especially since the blockchain cannot check the preimage before the swap happens. So I guess we need to ensure the preimage-length is limited by the same amount of bytes on all supported chains to avoid half atomic swaps.

Reply:
> You are right, SwapOfferTransaction and SwapCounterTransaction contains the same fields (probably they are implemented as the same transaction in BNS). But maybe they'll diverge at some point?

The transactions on the blockchain cannot diverge at any point. I used this for different intents when adding a pre-processing, but I agree that added confusion. As mentioned below, once you make them the same fields, please make them the same type. It will also make parsing more clear